provider/core/src/resource.rs

```diff
@@ -372,62 +373,113 @@ impl fmt::Display for ResourceOptions {
-            variant.write_to(sink)?;
-            initial = false;
-        }
-        if let Some(langid) = &self.langid {
-            if !initial {
-                sink.write_char('/')?;
-            }
-            langid.write_to(sink)?;
+        self.langid.write_to(sink)?;
+        if !self.keywords.is_empty() {
+            sink.write_str("-u-")?;
+            self.keywords.write_to(sink)?;
         }
         Ok(())
     }
 
     fn write_len(&self) -> LengthHint {
-        let mut length_hint = LengthHint::exact(0);
-        let mut initial = true;
-        if let Some(variant) = &self.variant {
-            length_hint += variant.write_len();
-            initial = false;
-        }
-        if let Some(langid) = &self.langid {
-            if !initial {
-                length_hint += 1;
-            }
-            length_hint += langid.write_len();
-        }
-        length_hint
+        self.langid.write_len()
+            + if !self.keywords.is_empty() { 3 } else { 0 }
+            + self.keywords.write_len()
     }
 }
 
 impl From<LanguageIdentifier> for ResourceOptions {
-    /// Create a ResourceOptions with the given language identifier and an empty variant field.
     fn from(langid: LanguageIdentifier) -> Self {
         Self {
-            langid: Some(langid),
-            variant: None,
+            langid,
+            keywords: unicode_ext::Keywords::new(),
         }
     }
 }
 
 impl From<Locale> for ResourceOptions {
-    /// Create a ResourceOptions with the given language identifier and an empty variant field.
     fn from(locale: Locale) -> Self {
+        // TODO(#1109): Implement proper vertical fallback
         Self {
-            langid: Some(locale.id),
-            variant: None,
+            langid: locale.id,
+            keywords: locale.extensions.unicode.keywords,
         }
     }
 }
 
 impl ResourceOptions {
+    /// TODO(#1109): Delete this function and use vertical fallback instead
+    pub fn temp_for_region(region: Option<Region>) -> Self {
+        Self {
+            langid: LanguageIdentifier::from(region),
+            keywords: unicode_ext::Keywords::new(),
+        }
+    }
+
     /// Returns whether this [`ResourceOptions`] has all empty fields (no components).
     pub fn is_empty(&self) -> bool {
         self == &Self::default()
     }
+
+    /// Returns whether the [`LanguageIdentifier`] associated with this request is `und`.
+    ///
+    /// Note that this only checks the language identifier; extension keywords may also be set.
+    /// To check the entire `ResourceOptions`, use [`ResourceOptions::is_empty()`].
+    pub fn is_langid_und(&self) -> bool {
+        self.langid == LanguageIdentifier::UND
+    }
+
+    /// Returns the [`LanguageIdentifier`] for this [`ResourceOptions`].
+    ///
+    /// If you have ownership over the `ResourceOptions`, use [`ResourceOptions::into_locale()`]
+    /// and then access the `id` field.
+    pub fn langid(&self) -> LanguageIdentifier {
```

Comment:
Why doesn't this borrow?

Reply:
(interpreting this as a `Question`)

Mainly because I wanted to stop exposing the internals of `ResourceOptions`. I've introduced the following functions:

1. This method, which constructs a new `LanguageIdentifier`, which *usually* doesn't allocate
2. Direct accessors for the language, script, and region subtags, which never allocate
3. `locale(self)` which moves out of `self` without allocating memory and without exposing the internals directly

I'm willing to discuss this and explore alternatives if you have suggestions.

Reply:
@Manishearth 

You made another comment but it's not showing up in this thread.

The main reason I don't want to return `&LanguageIdentifier` is because I don't want to expose the internals of ResourceOptions. For example, we may consider a change that stores the subtags flat with a single variant, or something else that is hyper optimized for vertical fallback. I expect that most things users would need to do with ResourceOptions should be methods directly on it. It's not the business of the user of this class that we're able to return a `&LanguageIdentifier`.